### PR TITLE
Direct store object

### PR DESCRIPTION
**Pull request by shans**
I'd appreciate comments on testing particularly, but obviously on implementation too.

With regards to testing, I have an idea for flow testing that I'll write up and send. The tests here are intended to be unit tests only - i.e. they specifically test particular properties of the store. The flow testing will be more about testing 'bulk properties' - they'll focus on the model/operation interfaces and assert that model properties are maintained independent of interleaving order.

### CODE REVIEW

File: src/runtime/storageNG/tests/store-test.ts [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    return true;
+  }
+  driver<Data>(storageKey: string, exists: Exists): Driver<Data> {
+    return new MockDriver<Data>(storageKey, exists);
+  }
+}
+
+describe('Store', async () => {
+
+  it(`will throw an exception if an appropriate driver can't be found`, async () => {
```
Comment (mykmartin) [inline src/runtime/storageNG/tests/store-test.ts]: This should exercise the willSupport logic when a driver has been registered but doesn't accept the given key.

File: src/runtime/storageNG/tests/store-test.ts [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    DriverFactory.register(new MockStorageDriverProvider());
+
+    const store = new Store('string', Exists.ShouldCreate, null, StorageMode.Direct, CRDTCount);
+    const activeStore = store.activate();
+
+    assert.equal(activeStore.constructor, DirectStore);
+  });
+
+  it('will propagate model updates from proxies to drivers', async () => {
+    DriverFactory.register(new MockStorageDriverProvider());
```
Comment (mykmartin) [inline src/runtime/storageNG/tests/store-test.ts]: This isn't hermetic; each call to register() adds another mock instance to the static list.

File: src/runtime/storageNG/tests/store-test.ts [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    return new Promise(async (resolve, reject) => {
+      const id = activeStore.on(proxyMessage => {
+        if (proxyMessage.type === ProxyMessageType.Operations) {
+          assert.isFalse(sentSyncRequest);
+          sentSyncRequest = true;
+          activeStore.onProxyMessage({type: ProxyMessageType.SyncRequest, id});
+          return true;
+        }
+        assert.isTrue(sentSyncRequest);
+        if (proxyMessage.type === ProxyMessageType.ModelUpdate) {
```
Comment (mykmartin) [inline src/runtime/storageNG/tests/store-test.ts]: Why is this required? Looks like this test only drives the ops proxy message type..

File: src/runtime/storageNG/drivers/driver-factory.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import {CRDTModel} from "../../crdt/crdt";
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/drivers/driver-factory.ts]: also .. unused, as it turns out.

File: src/runtime/storageNG/store.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {CRDTModel, CRDTTypeRecord, CRDTChange, ChangeType, CRDTError} from "../crdt/crdt";
+import {Type} from "../type";
+import {Exists, Driver, DriverFactory} from "./drivers/driver-factory";
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/store.ts]: Done.

File: src/runtime/storageNG/store.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+  constructor(storageKey: string, exists: Exists, type: Type, mode: StorageMode, modelConstructor: new () => CRDTModel<T>) {
+    this.storageKey = storageKey;
+    this.exists = exists;
+    this.type = type;
+    this.mode = mode;
+    this.modelConstructor = modelConstructor;
+  }
+  activate(): ActiveStore<T> {
+    const activeStore = new (this.constructors.get(this.mode))<T>(this.storageKey, this.exists, this.type, this.mode, this.modelConstructor);
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/store.ts]: Done.

File: src/runtime/storageNG/store.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    this.exists = Exists.ShouldExist;
+    return activeStore;
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/store.ts]: The pre-active store is the inactive record in the manifest, and could be activated multiple times e.g. when mapped from a running arc.

So I think caching is bad and it's expected to potentially have multiple active versions floating around at the same time.

File: src/runtime/storageNG/tests/store-test.ts (hunk 1 above)
Comment (shans) [reply src/runtime/storageNG/tests/store-test.ts]: That's outside the scope of unit tests of Store though. Perhaps I can test that separately when driver infrastructure is more fleshed out, or as a simple integration test, or something.

File: src/runtime/storageNG/tests/store-test.ts (hunk 2 above)
Comment (shans) [reply src/runtime/storageNG/tests/store-test.ts]: Added afterEach call to clean up.

File: src/runtime/storageNG/tests/store-test.ts (hunk 3 above)
Comment (shans) [reply src/runtime/storageNG/tests/store-test.ts]: The first if clause sends a sync request in response to operations being delivered. I wanted a non-empty model inside the store and it seemed nice to test this whole flow rather than just push one in.
